test/Microsoft.AspNet.Owin.Tests/OwinExtensionTests.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+
+namespace Microsoft.AspNet.Owin
+{
+    using AppFunc = Func<IDictionary<string, object>, Task>;
+    using CreateMiddleware = Func<
+          Func<IDictionary<string, object>, Task>,
+          Func<IDictionary<string, object>, Task>
+        >;
+    using AddMiddleware = Action<Func<
+          Func<IDictionary<string, object>, Task>,
+          Func<IDictionary<string, object>, Task>
+        >>;
+
+    public class OwinExtensionTests
+    {
+        static AppFunc notFound = async env => env["owin.ResponseStatusCode"] = 404;
+
+        [Fact]
+        public void OwinConfigureServiceProviderAddsServices()
+        {
+            IList<CreateMiddleware> list = new List<CreateMiddleware>();
+            AddMiddleware build = list.Add;
+            IServiceProvider serviceProvider = null;
+            FakeService fakeService = null;
+
+            var builder = build.UseBuilder(applicationBuilder =>
+            {
+                serviceProvider = applicationBuilder.ApplicationServices;
+                applicationBuilder.Run(async context =>
+                {
+                    fakeService = context.ApplicationServices.GetService<FakeService>();
+                });
+            }, new ServiceCollection().AddSingleton(new FakeService()).BuildServiceProvider());
+
+            list.Reverse().Aggregate(notFound, (next, middleware) => middleware(next)).Invoke(new Dictionary<string, object>());
+
+            Assert.NotNull(fakeService);
+            Assert.NotNull(serviceProvider?.GetService<FakeService>());
+        }
+
+        [Fact]
+        public void OwinDefaultNoServices()
+        {
+            IList<CreateMiddleware> list = new List<CreateMiddleware>();
+            AddMiddleware build = list.Add;
+            IServiceProvider serviceProvider = null;
+            FakeService fakeService = null;
+
+            var builder = build.UseBuilder(applicationBuilder =>
+            {
+                serviceProvider = applicationBuilder.ApplicationServices;
+                applicationBuilder.Run(async context =>
+                {
+                    fakeService = context.ApplicationServices.GetService<FakeService>();
+                });
+            });
+
+            list.Reverse().Aggregate(notFound, (next, middleware) => middleware(next)).Invoke(new Dictionary<string, object>());
+
+            Assert.Null(fakeService);
```

Comment:
Could you additionally set a bool inside the builder func and verify if it's true here? These values are just as likely to be null if the builder func never gets executed so we'd validate it actually was executed.